pose-detection/src/movenet/detector.ts

```diff
@@ -0,0 +1,371 @@
+   * details in the documentation of the ModelConfig.
+   *
+   * @param config ModelConfig dictionary that contains parameters for
+   * the MoveNet loading process. Please find more details of each parameters
+   * in the documentation of the ModelConfig interface.
+   */
+  static async load(modelConfig: MoveNetModelConfig = MOVENET_CONFIG):
+      Promise<PoseDetector> {
+    const config = validateModelConfig(modelConfig);
+    // TODO: add URLs here once we know them.
+    const defaultUrl = '';
+    const model: KeypointModel = new KeypointModel();
+    await model.load(config.modelUrl || defaultUrl);
+    return new MoveNetDetector(model, config);
+  }
+
+  /**
+   * Estimates poses for an image or video frame.
+   *
+   * This does standard ImageNet pre-processing before inferring through the
+   * model. The image should pixels should have values [0-255]. It returns a
+   * single pose or multiple poses based on the maxPose parameter from the
+   * `config`.
+   *
+   * @param image
+   * ImageData|HTMLImageElement|HTMLCanvasElement|HTMLVideoElement The input
+   * image to feed through the network.
+   *
+   * @param config
+   *       maxPoses: Optional. Max number of poses to estimate.
+   *       Only 1 pose is supported.
+   *
+   *       flipHorizontal: Optional. Default to false. When image data comes
+   *       from camera, the result has to flip horizontally.
+   *
+   * @return An array of `Pose`s.
+   */
+  async estimatePoses(
+      image: PoseDetectorInput,
+      estimationConfig:
+          MoveNetEstimationConfig = MOVENET_SINGLE_PERSON_ESTIMATION_CONFIG):
+      Promise<Pose[]> {
+    const config = validateEstimationConfig(estimationConfig);
+
+    if (image == null) {
+      return [];
+    }
+
+    this.maxPoses = config.maxPoses;
+
+    if (this.maxPoses > 1) {
+      throw new Error('Multi-person poses is not implemented yet.');
+    }
+
+    // Keep track of fps for one euro filter.
+    const now = performance.now();
+    if (this.previousFrameTime !== 0) {
+      this.frameTimeDiff =
+          0.02 * this.frameTimeDiff + 0.98 * (now - this.previousFrameTime);
+    }
+    this.previousFrameTime = now;
+
+    const imageTensor3D = toImageTensor(image);
+    const imageHeight = imageTensor3D.shape[0];
+    const imageWidth = imageTensor3D.shape[1];
+    const imageTensor4D = tf.expandDims(imageTensor3D, 0) as tf.Tensor4D;
+    imageTensor3D.dispose();
+
+    const [keypoints, ,] = await this.runInference(imageTensor4D, estimationConfig.minimumKeypointScore);
+
+    for (let i = 0; i < keypoints.length; ++i) {
+      keypoints[i].y *= imageHeight;
+      keypoints[i].x *= imageWidth;
+    }
+
+    const poses: Pose[] = [];
+    poses[0] = {'keypoints': keypoints};
+
+    return poses;
+  }
+
+  determineCropRegion(
+      keypoints: Keypoint[], webcamHeight: number, webcamWidth: number,
+      minimumKeypointScore: number) {
+    const keypointIndices: {[index: string]: number} = {
```

Comment:
i noticed that constants.ts also defines all the coco keypoints (although with different names). does it make sense to use that, or do you think it's better to keep it here in case we eventually want to use a different set of keypoints?